stdlib/Makefile

```diff
@@ -13,66 +13,236 @@
+	  camlheader_ur \
+	  "$(INSTALL_LIBDIR)"
+	cp target_camlheader "$(INSTALL_LIBDIR)/camlheader"
+
+ifeq "$(RUNTIMED)" "runtimed"
+install::
+	cp target_camlheaderd $(INSTALL_LIBDIR)
+endif
+
+ifeq "$(RUNTIMEI)" "true"
+install::
+	cp target_camlheaderi $(INSTALL_LIBDIR)
+endif
+
+.PHONY: installopt
 installopt: installopt-default installopt-$(PROFILING)
 
+.PHONY: installopt-default
 installopt-default:
-	cp stdlib.cmxa stdlib.a std_exit.o *.cmx "$(INSTALL_LIBDIR)"
-	cd "$(INSTALL_LIBDIR)"; $(RANLIB) stdlib.a
+	cp stdlib.cmxa stdlib.$(A) std_exit.$(O) *.cmx "$(INSTALL_LIBDIR)"
+	cd "$(INSTALL_LIBDIR)"; $(RANLIB) stdlib.$(A)
 
+.PHONY: installopt-noprof
 installopt-noprof:
-	rm -f "$(INSTALL_LIBDIR)/stdlib.p.cmxa"; \
-	  ln -s stdlib.cmxa "$(INSTALL_LIBDIR)/stdlib.p.cmxa"
-	rm -f "$(INSTALL_LIBDIR)/stdlib.p.a"; \
-	  ln -s stdlib.a "$(INSTALL_LIBDIR)/stdlib.p.a"
-	rm -f "$(INSTALL_LIBDIR)/std_exit.p.cmx"; \
-	  ln -s std_exit.cmx "$(INSTALL_LIBDIR)/std_exit.p.cmx"
-	rm -f "$(INSTALL_LIBDIR)/std_exit.p.o"; \
-	  ln -s std_exit.o "$(INSTALL_LIBDIR)/std_exit.p.o"
+	ln -sf stdlib.cmxa "$(INSTALL_LIBDIR)/stdlib.p.cmxa"
+	ln -sf stdlib.$(A) "$(INSTALL_LIBDIR)/stdlib.p.$(A)"
+	ln -sf std_exit.cmx "$(INSTALL_LIBDIR)/std_exit.p.cmx"
+	ln -sf std_exit.$(O) "$(INSTALL_LIBDIR)/std_exit.p.$(O)"
 
+.PHONY: installopt-prof
 installopt-prof:
-	cp stdlib.p.cmxa stdlib.p.a std_exit.p.cmx std_exit.p.o \
+	cp stdlib.p.cmxa stdlib.p.$(A) std_exit.p.cmx std_exit.p.$(O) \
 	  "$(INSTALL_LIBDIR)"
-	cd "$(INSTALL_LIBDIR)"; $(RANLIB) stdlib.p.a
+	cd "$(INSTALL_LIBDIR)"; $(RANLIB) stdlib.p.$(A)
+
+ifeq "$(UNIX_OR_WIN32)" "unix"
+HEADERPROGRAM = header.c
+else # Windows
+HEADERPROGRAM = headernt.c
+endif
+
+CAMLHEADERS =\
+  camlheader target_camlheader camlheader_ur \
+  camlheaderd target_camlheaderd \
+  camlheaderi target_camlheaderi
+
+ifeq "$(HASHBANGSCRIPTS)" "true"
+$(CAMLHEADERS): ../config/Makefile
+	for suff in '' d i; do \
+	  echo '#!$(BINDIR)/ocamlrun'$$suff > camlheader$$suff && \
+	  echo '#!$(TARGET_BINDIR)/ocamlrun'$$suff >target_camlheader$$suff; \
+	done && \
+	echo '#!' | tr -d '\012' > camlheader_ur;
+else # Hashbang scripts not supported
+
+$(CAMLHEADERS): $(HEADERPROGRAM) ../config/Makefile
+
+ifeq "$(UNIX_OR_WIN32)" "unix"
+$(CAMLHEADERS):
+	for suff in '' d i; do \
+	  $(BYTECC) $(BYTECCCOMPOPTS) $(BYTECCLINKOPTS) \
+	            -DRUNTIME_NAME='"$(BINDIR)/ocamlrun'$$suff'"' \
+	            header.c -o tmpheader$(EXE) && \
+	  strip tmpheader$(EXE) && \
+	  mv tmpheader$(EXE) camlheader$$suff && \
+	  $(BYTECC) $(BYTECCCOMPOPTS) $(BYTECCLINKOPTS) \
+	            -DRUNTIME_NAME='"$(TARGET_BINDIR)/ocamlrun'$$suff'"' \
+	            header.c -o tmpheader$(EXE) && \
+	  strip tmpheader$(EXE) && \
+	  mv tmpheader$(EXE) target_camlheader$$suff; \
+	done && \
+	cp camlheader camlheader_ur
+
+else # Windows
+
+# TODO: see whether there is a way to further merge the rules below
+# with those above
+
```

Comment:
Only with tremendous care - I think more than one of the lines below has been the subject of its own Mantis PR!